### PR TITLE
fix(runtime): bound write_file execution-failure observations

### PR DESCRIPTION
**Pull request by kibabsquirrel**
## Summary

Related to [#3114 ](https://github.com/bytedance/deer-flow/issues/3114)This PR bounds `write_file` execution-failure observations.

The runtime already truncates tool exceptions that are raised into `ToolErrorHandlingMiddleware`, but `write_file_tool()` currently catches its own exceptions and returns error strings directly. That bypasses the framework-level exception truncation path, so oversized write failures can still feed large observations back into later model context.

This patch adds bounded formatting for that tool-local catch/return path while preserving the current tool contract and local path sanitization behavior.

## Key Decisions

- keep the fix local to `write_file` instead of changing the tool to re-raise into middleware
- preserve `_sanitize_error()` path masking for local sandbox output
- apply bounded formatting across `SandboxError`, `PermissionError`, `IsADirectoryError`, `OSError`, and generic `Exception` paths
- keep this as a containment fix for the execution-failure surface; broader payload/history inflation should be handled separately

## Tests

Ran:

```bash
uv run pytest tests/test_sandbox_tools_security.py -v
```

Added regression coverage for:

- large `OSError` is bounded and still preserves diagnostic context
- local sandbox host paths are still masked
- short `OSError` remains readable and untruncated
- large `SandboxError` is bounded
- `PermissionError`, `IsADirectoryError`, and generic `Exception` all follow the bounded formatting path without regressing short-error readability

### CODE REVIEW

**Comment by WillemJiang** [comment]
@kibabsquirrel, thanks for your contribution. Please take a look at the review comments.

File: backend/tests/test_sandbox_tools_security.py [hunk 1]
```diff
@@ -1140,6 +1141,145 @@ def append_worker() -> None:
...
+        path="/mnt/user-data/workspace/output.txt",
+        content="report body",
+    )
+
+    assert len(result) <= 2000
+    assert "Error: Failed to write file '/mnt/user-data/workspace/output.txt':" in result
+    assert "/tmp/deer-flow/threads/t1/user-data/workspace" not in result
+    assert "/mnt/user-data/workspace/nested/output.txt" in result
+    assert "remote tail marker" in result
+    assert "[write_file error truncated:" in result
```
[the remark this answers is not in the document]
Comment (kibabsquirrel) [reply backend/tests/test_sandbox_tools_security.py:1177]: The hardcoded 2000 is intentional — it pins the documented default contract. If _DEFAULT_WRITE_FILE_ERROR_MAX_CHARS is changed, the test failure forces a deliberate review of whether the new value still satisfies the BUG-003 containment requirement. Resolving as-is.

File: backend/packages/harness/deerflow/sandbox/tools.py [hunk 2]
```diff
@@ -1525,15 +1563,25 @@ def write_file_tool(
...
+            e,
+            runtime,
+            message=f"Error: Path is a directory, not a file: {requested_path}",
+        )
     except OSError as e:
-        return f"Error: Failed to write file '{requested_path}': {_sanitize_error(e, runtime)}"
+        return _format_write_file_error(requested_path, e, runtime)
     except Exception as e:
-        return f"Error: Unexpected error writing file: {_sanitize_error(e, runtime)}"
+        return _format_write_file_error(requested_path, e, runtime)
```
[the remark this answers is not in the document]
Comment (kibabsquirrel) [reply backend/packages/harness/deerflow/sandbox/tools.py:1579]: Fixed in 2908255b: `PermissionError` and `IsADirectoryError` now
return self-contained messages (no redundant `: PermissionError:
permission denied` suffix). Other branches keep the
`: {ExceptionType}: {detail}` format intentionally — the type
prefix is a stable error class the model can branch on.

**Comment by kibabsquirrel** [comment]
@WillemJiang  Addressed Copilot review in 2908255b. The original
single-formatter shape kept truncation and sanitization uniform,
but for `PermissionError` / `IsADirectoryError` the type prefix
duplicated the already-self-describing message — those two now
return directly, while the other branches keep the type-prefix
format intentionally for stable, machine-readable error classes.
Ready for another look.

**Comment by ShenAC-SAC** [comment]
@kibabsquirrel I found one edge case in the new `write_file` error handling path.

`requested_path` is assigned after sandbox initialization:

```python
sandbox = ensure_sandbox_initialized(runtime)
ensure_thread_directories_exist(runtime)
requested_path = path
```

but the new `except SandboxError` branch uses it:

```python
except SandboxError as e:
    return _format_write_file_error(requested_path, e, runtime)
```

So if `ensure_sandbox_initialized()` or `ensure_thread_directories_exist()` raises `SandboxRuntimeError` / `SandboxNotFoundError`, `write_file_tool` can raise `UnboundLocalError` before returning the bounded tool error.

I reproduced this locally by monkeypatching `ensure_sandbox_initialized()` to raise `SandboxRuntimeError("sandbox missing")`.

The fix should be small: initialize `requested_path = path` before any call that can raise, and add a regression test for the sandbox-initialization failure path so this stays covered.

**Comment by kibabsquirrel** [comment]
@ShenAC-SAC Thanks for catching this edge case. I moved `requested_path = path` before sandbox setup so sandbox initialization failures can still return the bounded `write_file` error instead of masking the original failure with `UnboundLocalError`.

I also added a regression test that monkeypatches sandbox initialization to fail and verifies the tool returns the bounded error path. I had focused the first pass on bounding/sanitizing write errors after execution, and I’ll be more careful to audit early failure paths and variables referenced from shared exception handlers in similar changes.

**Comment by WillemJiang** [comment]
@kibabsquirrel, could you please fix the lint error?

**Comment by kibabsquirrel** [comment]
@WillemJiang You're right — thanks for catching it.

I'll be more careful about  full lint checks on follow-up updates.

**Review by WillemJiang — APPROVED** [review]
